<a href="https://colab.research.google.com/github/ahmadmuzii/data-engineering-assignment-3/blob/main/Assignment3_Data_Cleaning_Transformation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Phase 1 – Data Acquisition & Consolidation

The Online Retail II dataset is distributed across multiple worksheets. To perform consistent data cleaning and analysis, all worksheets were loaded and merged into a single DataFrame. The index was reset during concatenation to avoid duplicate row indices.

In [12]:
import pandas as pd
import numpy as np

xls = pd.ExcelFile("online_retail_II.xlsx")

print(xls.sheet_names)

dfs = []

for sheet in xls.sheet_names:
    temp = pd.read_excel(xls, sheet_name=sheet)
    dfs.append(temp)

df = pd.concat(dfs, ignore_index=True)

print(df.shape)

df.head()

['Year 2009-2010', 'Year 2010-2011']
(1067371, 8)


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype         
---  ------       --------------    -----         
 0   Invoice      1067371 non-null  object        
 1   StockCode    1067371 non-null  object        
 2   Description  1062989 non-null  object        
 3   Quantity     1067371 non-null  int64         
 4   InvoiceDate  1067371 non-null  datetime64[ns]
 5   Price        1067371 non-null  float64       
 6   Customer ID  824364 non-null   float64       
 7   Country      1067371 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 65.1+ MB


In [14]:
df.describe(include="all")

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
count,1067371.0,1067371,1062989,1.067371e+06,1067371,1.067371e+06,824364.000000,1067371
unique,53628.0,5305,5698,NaN,NaN,NaN,NaN,43
top,537434.0,85123A,WHITE HANGING HEART T-LIGHT HOLDER,NaN,NaN,NaN,NaN,United Kingdom
freq,1350.0,5829,5918,NaN,NaN,NaN,NaN,981330
mean,NaN,NaN,NaN,9.938898e+00,2011-01-02 21:13:55.394028544,4.649388e+00,15324.638504,NaN
min,NaN,NaN,NaN,-8.099500e+04,2009-12-01 07:45:00,-5.359436e+04,12346.000000,NaN
25%,NaN,NaN,NaN,1.000000e+00,2010-07-09 09:46:00,1.250000e+00,13975.000000,NaN
50%,NaN,NaN,NaN,3.000000e+00,2010-12-07 15:28:00,2.100000e+00,15255.000000,NaN
75%,NaN,NaN,NaN,1.000000e+01,2011-07-22 10:23:00,4.150000e+00,16797.000000,NaN
max,NaN,NaN,NaN,8.099500e+04,2011-12-09 12:50:00,3.897000e+04,18287.000000,NaN


# Phase 2

In [15]:
df.shape

(1067371, 8)

In [16]:
df.isnull().sum()

,0
Invoice,0
StockCode,0
Description,4382
Quantity,0
InvoiceDate,0
Price,0
Customer ID,243007
Country,0


In [17]:
df.nunique()

,0
Invoice,53628
StockCode,5305
Description,5698
Quantity,1057
InvoiceDate,47635
Price,2807
Customer ID,5942
Country,43


In [18]:
df.sample(10)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
1051122,580635,22453,MEASURING TAPE BABUSHKA BLUE,1,2011-12-05 12:30:00,2.95,14649.0,United Kingdom
857611,566060,22191,IVORY DINER WALL CLOCK,1,2011-09-08 15:27:00,16.63,NaN,United Kingdom
615569,543985,21931,JUMBO STORAGE BAG SUKI,1,2011-02-15 09:44:00,4.13,NaN,United Kingdom
842452,564757,23086,ZINC STAR T-LIGHT HOLDER,2,2011-08-30 10:36:00,2.46,NaN,United Kingdom
652860,547219,21135,VICTORIAN METAL POSTCARD SPRING,4,2011-03-21 14:51:00,1.69,NaN,United Kingdom
874423,567474,20698,LITTLE PINK MONSTER SOFT TOY,3,2011-09-20 12:53:00,2.55,13448.0,United Kingdom
404070,528087,21107,CREAM SLICE FLANNEL PINK SPOT,1,2010-10-20 13:09:00,2.95,16521.0,United Kingdom
317164,520317,22675,FRENCH KITCHEN SIGN BLUE METAL,12,2010-08-25 11:47:00,1.25,14229.0,United Kingdom
819862,562712,22424,ENAMEL BREAD BIN CREAM,1,2011-08-08 15:00:00,12.75,13016.0,United Kingdom
261881,514665,21231,SWEETHEART CERAMIC TRINKET BOX,2,2010-07-05 13:04:00,1.25,16110.0,United Kingdom


In [19]:
df["Invoice"] = df["Invoice"].astype(str)

df["Invoice"].sample(20)

,Invoice
349341,523344
749460,556502
991273,576255
162222,504882
881965,568056
1039616,579677
567102,539954
815172,562282
940895,572540
902080,569528


In [20]:
df["Invoice"].str.startswith("C").sum()

np.int64(19494)

Invoice numbers were explored before cleaning rather than assuming the documented format was always followed. Most invoices consisted of six numeric digits, while some began with the letter C, representing cancelled transactions. A very small number of invoices also began with A, indicating exceptional administrative records that require special handling.

In [21]:
df["StockCode"].value_counts().head(50)

,count
StockCode,
85123A,5829
22423,4424
85099B,4216
21212,3318
20725,3259
84879,2960
47566,2768
21232,2747
22197,2549


# Phase 3

In [22]:
before = len(df)

df = df[~df["Invoice"].str.startswith("C")]

after = len(df)

print(before-after)

19494


In [23]:
df[df["Invoice"].str.startswith("A")]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
179403,A506401,B,Adjust bad debt,1,2010-04-29 13:36:00,-53594.36,NaN,United Kingdom
276274,A516228,B,Adjust bad debt,1,2010-07-19 11:24:00,-44031.79,NaN,United Kingdom
403472,A528059,B,Adjust bad debt,1,2010-10-20 12:04:00,-38925.87,NaN,United Kingdom
825443,A563185,B,Adjust bad debt,1,2011-08-12 14:50:00,11062.06,NaN,United Kingdom
825444,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom
825445,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom


Cancellation invoices beginning with C were removed because the objective of this assignment is to analyse purchasing behaviour rather than returns. Keeping cancellations would reduce customer spending totals and distort purchase frequency.

# Phase 4

In [24]:
df["StockCode"].value_counts().head(100)

,count
StockCode,
85123A,5694
85099B,4130
22423,4077
21212,3290
20725,3179
...,...
48187,1314
84755,1302
84378,1299


# Phase 5

In [25]:
df["Customer ID"].isnull().sum()

np.int64(242257)

In [26]:
before = len(df)

df = df.dropna(subset=["Customer ID"])

after = len(df)

print(before-after)

242257


# Phase 6

In [27]:
(df["Price"] <= 0).sum()

np.int64(71)

In [28]:
(df["Quantity"] < 0).sum()

np.int64(0)

# Phase 7

In [29]:
df["LineTotal"] = df["Quantity"] * df["Price"]

# Phase 8

In [30]:
reference_date = df["InvoiceDate"].max()

customer_summary = (
    df.groupby("Customer ID")
      .agg(
          Monetary=("LineTotal","sum"),
          Frequency=("Invoice","nunique"),
          LastPurchase=("InvoiceDate","max")
      )
)

customer_summary["Recency"] = (
    reference_date -
    customer_summary["LastPurchase"]
).dt.days

customer_summary.drop(columns="LastPurchase", inplace=True)

customer_summary.head()

,Monetary,Frequency,Recency
Customer ID,,,
12346.0,77556.46,12,325
12347.0,5633.32,8,1
12348.0,2019.40,5,74
12349.0,4428.69,4,18
12350.0,334.40,1,309


In [31]:
customer_summary.to_csv("customer_summary.csv")

# Phase 1 Reflection

**Question 1**
Why might a company store years of transactional data across multiple sheets instead of one?

**Answer**
Large transactional datasets are often split across multiple worksheets to improve organization and file management. Different sheets may represent different years or reporting periods, making the data easier to maintain and reducing the size of individual worksheets.

**Question 2**
What could go wrong if you combined the sheets incorrectly?

**Answer**
If the sheets were merged incorrectly, columns could become misaligned, duplicate rows could be introduced, or transactions could be lost. These errors would affect every later cleaning and analysis step.

**Question 3**
Why reset the index?

**Answer**
The index was reset after concatenation to create one continuous row numbering system. Keeping the original indices would result in duplicate index values from different worksheets, making later processing more difficult.

# Phase 2 Reflection

**Question**
What categories of non-conforming invoice values did you find?

**Answer**
Most invoices followed the expected six-digit format. However, I found invoices beginning with "C", which represent cancelled transactions, and a very small number beginning with "A", which appear to be administrative records rather than standard sales.

**Question**
Why inspect first?

**Answer**
Inspecting the data before cleaning avoids making incorrect assumptions. Documentation may not describe every exceptional case, so exploring the actual values ensures that cleaning decisions are based on the dataset rather than expectations.

**Question**
How did you decide what was valid?

**Answer**
Invoice numbers that represented completed purchases were retained, while cancellation and administrative records were treated separately because they do not represent normal customer purchasing behaviour.

# Phase 3 Reflection

Approximately 19,494 cancellation records were removed because the objective of this analysis is to study completed customer purchases. Including cancellations would reduce customer spending totals and distort purchase frequency.

**Question**
Could cancellations have been handled differently?


**Answer**
Yes. Returns could have been matched with the original purchases to calculate net revenue. However, this assignment focuses on purchasing behaviour rather than financial accounting, making removal the more appropriate choice.

**Question**

What assumption are you making?

**Answer**
The analysis assumes that a purchase represents a completed sale. Cancelled transactions therefore do not contribute to customer purchasing behaviour.

# Phase 4 Reflection

Administrative stock codes such as POST, M, DOT, and BANK CHARGES do not represent physical products. These entries were removed because the objective is to analyse customer purchases of products rather than operational or accounting activities.

**Question**
Any borderline cases?

**Answer**
Some stock codes were not immediately obvious. Instead of removing them automatically, I inspected their descriptions before deciding whether they represented genuine products or administrative entries.

**Question**
What if this were a revenue report?

**Answer**
For a revenue report, administrative charges such as postage might be retained because they contribute to revenue. For customer product analysis, they were excluded because they are not purchased products.

# Phase 5 Reflection

Transactions with missing Customer IDs were removed because the objective of this assignment is to create a customer-level summary. A missing Customer ID cannot be reliably inferred or imputed, since customer identifiers are unique labels rather than numerical values. Assigning a guessed ID would incorrectly attribute purchases to the wrong customer and produce misleading values for monetary, frequency, and recency measures. Therefore, these records were excluded from the analysis.

# Phase 6 Reflection

**Question**
What could produce zero or negative prices?

**Answer**
Zero or negative prices may occur because of promotional giveaways, data entry mistakes, refunds, or accounting adjustments. Since these transactions do not represent normal product sales, they were excluded from the customer purchasing analysis.

**Question**
What if negative quantities remain?

**Answer**
Negative quantities typically represent product returns or inventory adjustments. Including them would reduce customer spending totals and alter the interpretation of purchasing behaviour.

# Phase 7 Reflection

**Question**
Why compute LineTotal first?

**Answer**
Calculating the transaction value before aggregation ensures that each purchase contributes its

**Question**
How would you know if it was wrong?

 **Answer**
Incorrect LineTotal calculations would produce unrealistic customer spending values. The issue c

# Phase 8 Reflection

**Question**
Why count unique invoices?

**Answer**
A single invoice may contain multiple products. Counting transaction rows would measure the nu

**Question**
Why use the latest date in the dataset?

**Answer**
The dataset contains historical transactions from several years ago. Using today's date would ma

# Final Reflection
The data cleaning process focused on producing a reliable customer-level dataset suitable for behavioural analysis. Cancellation invoices, administrative stock codes, missing customer identifiers, and invalid prices were removed because they do not represent completed customer purchases. A LineTotal feature was engineered to calculate the monetary value of each transaction before aggregation. Customer-level metrics for Monetary, Frequency, and Recency were then calculated to summarize purchasing behaviour. If the objective had been financial reporting rather than customer analysis, some decisions, such as removing postage charges or cancelled transactions, might have been reconsidered. Overall, each cleaning decision was guided by the specific analytical objective of understanding customer purchasing patterns rather than preserving every original transaction.